### ISSUE TITLE
In the implementation of Redis persistence, the GetKey method does not use the grainType parameter

### ISSUE THREAD

**Issue by gusuchengnan** [dotnet/orleans#8485]
https://github.com/dotnet/orleans/blame/bfc50a3828cb5222dfb5b9df154784b7ee8474c4/src/Redis/Orleans.Persistence.Redis/Storage/RedisGrainStorage.cs#L106

If multiple T-type data needs to be persisted internally in Grain, such as T t1 and T t2, it may not be supported.

In addition, for lists or hash type data that need to be persisted through Redis, the current implementation is to serialize and deserialize it as a complete object. When the amount of data is large, the size of the object will be relatively large, leading to a decrease in the read and write performance of persistent data. Is there any good solution to this problem?

**Comment by ReubenBond** [comment]
Good point, we should fix this and incorporate the 'grainType' parameter into the key.

For the second point, it's best to keep frequently updated grain data relatively small when using grain persistence. If you need something more complex, I recommend using the DB directly or using a different persistence abstraction

**Comment by gusuchengnan** [comment]
Okay, got it. Thanks very much.

**Comment by idanbenyehoshua** [comment]
@ReubenBond
In case of key changes, we must a migration am I right?
If not, after Nuget upgrade, the "old" key will by forever unused with all data. The state will be lost!

**Comment by ReubenBond** [comment]
@idanbenyehoshua, please take a look at https://github.com/dotnet/orleans/pull/8487 and tell me whether you think the hook to configure the storage key will satisfy your needs.

**Comment by idanbenyehoshua** [comment]
Excellent! Looks good.
Thank you!
** Just let you know that this package is used in production in my service. Works as expected.
